activesupport/lib/active_support/core_ext/enumerable.rb

```diff
@@ -17,12 +18,43 @@ module Enumerable
   # The default sum of an empty list is zero. You can override this default:
   #
   #  [].sum(Payment.new(0)) { |i| i.amount } # => Payment.new(0)
-  def sum(identity = nil, &block)
-    if block_given?
-      map(&block).sum(identity)
-    else
-      sum = identity ? inject(identity, :+) : inject(:+)
-      sum || identity || 0
+  if Enumerable.instance_methods(false).include?(:sum) && !(%w[a].sum rescue false)
+      alias :orig_enum_sum :sum
+
+      def sum(identity = nil, &block) #:nodoc:
+        if identity
+          orig_enum_sum(identity, &block)
+        elsif identity = _identity_for_sum
+          orig_enum_sum(identity, &block)
+        else
+          if block_given?
+            map(&block).sum(identity)
+          else
+            sum = identity ? inject(identity, :+) : inject(:+)
+            sum || identity || 0
+          end
+        end
+      end
+
+      private def _identity_for_sum #:nodoc:
+        return 0 if first.is_a?(Numeric)
+        ActiveSupport::Deprecation.warn(<<-MSG.squish)
+          Support for calling #sum on a non-numeric enumerable without an identity argument,
+          is deprecated and will soon be removed.
```

Comment:
Yeah, though the warning belongs in `#sum` itself. Since we don't want to warn if the elements are numeric, we could split that conditional out:
```ruby
if identity
  …
elsif first.is_a?(Numeric)
  …
elsif identity = _identity_for_sum
  …
elsif block_given?
  …
else
  …
end
```